src/hooks/useModal.tsx

```diff
@@ -0,0 +1,68 @@
+import React from "react";
+import useFocusTrap from "@charlietango/use-focus-trap";
+
+export function useModal(initialState = false) {
+  const [isOpen, setOpen] = React.useState(initialState);
+  const close = React.useCallback(() => setOpen(false), []);
+  const open = React.useCallback(() => setOpen(true), []);
+  const ref = React.useRef<HTMLElement>();
+  const focusTrapRef = useFocusTrap(isOpen);
+  const modalRef = React.useCallback(
+    (node: HTMLElement | null) => {
+      focusTrapRef(node);
+      if (node) {
+        ref.current = node;
+      }
+    },
+    [focusTrapRef]
+  );
+
+  // close on esc key
+  React.useEffect(() => {
+    if (!isOpen) {
+      return;
+    }
+    function handler(event: KeyboardEvent) {
+      if (event.key !== "Escape") {
+        return;
+      }
+      close();
+    }
+    window.addEventListener("keydown", handler);
+    return () => window.removeEventListener("keydown", handler);
+  }, [close, isOpen]);
```

Comment:
i could see a key listener hook being abstracted into its own utility, that would be cool. not something you need for this pr though